### ISSUE TITLE
vacation.pl stuck, no logs or debug - debian stretch - migration

### ISSUE THREAD

**Issue by syntaxerrormmm** [postfixadmin/postfixadmin#160]
**Product**: postfixadmin from git master
**Component**: VIRTUAL_VACATION/vacation.pl, version 4.2
**Platform**: new custom mail server, testing virtual machine
**Operating system**: Debian 9 Stretch, amd64, up-to-date
**Other software**: postfix (3.1.8-0+deb9u1 from repo), dovecot (2.2.27-3+deb9u2 from repo), mariadb (10.1.26-0+deb9u1 from repo)
**Priority**: P3
**Severity**: blocker (cannot use functionality)

**Description**: Activating vacation from the web interface (and configuring it from config.inc.php) then sending an email to the vacated address, no vacation mail is received. The missing functionality is due to the vacation.pl script not working and stucking.

**Detailed description**: Checking within SMTP log:
```
Mar 29 10:46:12 mail2 postfix/pipe[10241]: C6FEA2015D: to=<sender#domain.ext@autoreply.domain.ext>, orig_to=<sender@domain.ext>, relay=vacation, delay=0.48, delays=0.06/0.02/0/0.4, dsn=2.0.0, status=sent (delivered via vacation service (Log4perl: Seems like no initialization happened. Forgot to call init()?))
```

Running the script by hand, we got:
```
# sudo -u vacation /var/spool/vacation/vacation.pl -f sender@domain.ext -- sender@domain.ext
Log4perl: Seems like no initialization happened. Forgot to call init()?
```
And then the program is stuck.

Again, running:
```
# sudo -u vacation /var/spool/vacation/vacation.pl -t yes -f sender@domain.ext -- sender@domain.ext
2018/03/29 12:34:13 DEBUG> /var/spool/vacation/vacation_new.pl:581 main:: - Script argument SMTP recipient is : 'sender@domain.ext' and smtp_sender : 'sender@domain.ext'
```
No logs are provided (/var/log/vacation.log is empty, I touched it and assigned correct permisssions to be sure), no other debug outputs are provided.

Database was migrated from an old installation (obviously working). I have tried to port the old script (v4.0) on the new machine with no avail.

**Steps to reproduce**
1) `git clone https://github.com/postfixadmin/postfixadmin.git`
2) Configure *config.inc.php*
3) Configure vacation message by Web interface for sender@domain.ext
3) Configure vacation as by INSTALL.md in VIRTUAL_VACATION folder (stretch doesn't have *libgetopt-argvfile-perl*, installed *libgetopt-complete-perl*)
4) Configure *vacation.pl* with db variables
5) Try launch commands stated before or send an email to sender@domain.ext

**Actual results**: Cannot send vacation emails, vacation.pl is stuck with a non-blocking error.
**Expected results**: Able to send vacation emails, vacation.pl should run without error.

**Additional information**: I have no working knowledge of perl to be able to debug this one. I hope someone will easily find out what the problem is. If you need configuration files, please ask.

**Comment by syntaxerrormmm** [comment]
> And then the program is stuck.

I mean that I had to kill the script with Ctrl+C to end the loop.

**Comment by DavidGoodwin** [comment]
I don't have a debian stretch installation to easily verify/test with. Does anything show up if you try to strace it? Does Postfix log anything after you Ctrl+C vacation.pl?

**Comment by syntaxerrormmm** [comment]
> Does anything show up if you try to strace it? Does Postfix log anything after you Ctrl+C vacation.pl?

Nope, none at all. Still, the first run with `strace` seems like the script is waiting for some input, so probably it is expected to run in pipe with something else. I would do some further tests with `strace` and report back.

**Comment by DavidGoodwin** [comment]
You have something like : 
```
vacation    unix  -       n       n       -       -       pipe
  flags=Rq user=whoever argv=/some/path/to/vacation.pl -f ${sender} -- ${recipient}
```

in /etc/postfix/master.cf ?

**Comment by syntaxerrormmm** [comment]
> You have something like in /etc/postfix/master.cf ?

Yup:
```
# tail -n3 /etc/postfix/master.cf
# Vacation
vacation  unix	-	n	n	-	-	pipe
  flags=Rq user=vacation argv=/var/spool/vacation/vacation.pl -f ${sender} -- ${recipient}
```

**Comment by DavidGoodwin** [comment]
Ah, sorry, I should have noticed your thing above ... yes, it definitely has to have something piped into it.

I normally pipe one of the files in tests/ into it ... ( see : https://github.com/postfixadmin/postfixadmin/tree/master/VIRTUAL_VACATION/tests )

i.e. cat tests/spam.txt | perl vacation.pl -t yes -f something@example.com -- random\#example.com@autoreply.domain

(There's a test.sh script in there too which might help).

**Comment by syntaxerrormmm** [comment]
Ok, I'll leave here the strace log attached. Produced as:
```
$ echo "Hello World!" | strace -o /tmp/strace_vacation.log perl vacation.pl -f sender@domain.ext -- sender@domain.ext
Log4perl: Seems like no initialization happened. Forgot to call init()?
```
**CAUTION**: obviously sender address is rewritten in github.

I am a noobie at `strace`, but to me some remarkable lines are these:
```
socket(AF_UNIX, SOCK_STREAM, 0)         = 3
connect(3, {sa_family=AF_UNIX, sun_path="/var/run/mysqld/mysqld.sock"}, 110) = 0
setsockopt(3, SOL_IP, IP_TOS, [8], 4)   = -1 EOPNOTSUPP (Operation not supported)
setsockopt(3, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0
recvfrom(3, "b\0\0\0\n5.5.5-10.1.26-MariaDB-0+deb"..., 16384, 0, NULL, NULL) = 102
brk(0x55befcdce000)                     = 0x55befcdce000
```

So it presumes to have the db ready with socket, which shouldn't be the case (from vacation.pl itself):
```
# db_type - uncomment one of these
#our $db_type = 'Pg';
our $db_type = 'mysql';

# leave empty for connection via UNIX socket
our $db_host = 'localhost';

# connection details
our $db_username = 'postfix';
our $db_password = 'xxxxxxxx';
our $db_name     = 'postfix';
```

[strace_vacation.log](https://github.com/postfixadmin/postfixadmin/files/1860448/strace_vacation.log)

**Comment by DavidGoodwin** [comment]
Try 127.0.0.1 if you want to force it to use tcp; I think MySQL will use a socket if you say 'localhost'.

Having said that, I think it queried MySQL fine (after all you have the 'recvfrom' line in strace).

vacation.pl should have some sort of debug output / setting - does turning that on help?

**Comment by syntaxerrormmm** [comment]
> Try 127.0.0.1 if you want to force it to use tcp; I think MySQL will use a socket if you say 'localhost'.

Doesn't change too much, from what I have seen.

> vacation.pl should have some sort of debug output / setting - does turning that on help?

Yes, that helped a lot.
I watched your tests and now I understand. The *vacation.pl* script does not accept any text in stdin, it needs a valid mail message (to take *from*, *to*, *subject* fields and so on). Also, vacation messages will not be sent if sender and recipient are the same address.

So I create a new test user on the system and tried to send an email to the previous vacated user and it works completely fine.

In `mail.log` there's always the line:
```
Log4perl: Seems like no initialization happened. Forgot to call init()?
```
but, as I stated in the issue opening it doesn't seem blocking (and vacation.log is empty, anyways).
